server/src/main/java/dev/slimevr/vr/trackers/IMUTracker.java

```diff
@@ -362,6 +431,136 @@ private void fixYaw(Quaternion reference) {
 		yawFix.set(sensorRotation).inverseLocal().multLocal(targetRotation);
 	}
 
+	private void calibrateMag() {
+		if (magCalibrationStatus >= CalibrationAccuracy.HIGH.status) {
+			magnetometerCalibrated = true;
+			// During calibration set correction to match magnetometer readings
+			// TODO : Correct only yaw
+			correction.set(rotQuaternion).inverseLocal().multLocal(rotMagQuaternion);
+		}
+	}
+
+	/**
+	 * Calculates 1 since last reset and store the data related to it in
+	 * driftQuat, timeAtLastReset and timeForLastReset
+	 */
+	synchronized public void calculateDrift(Quaternion beforeQuat) {
+		if (compensateDrift && allowDriftCompensation) {
+			Quaternion rotQuat = new Quaternion();
+			getUnfilteredRotation(rotQuat);
+			if (
+				driftSince > 0
+					&& System.currentTimeMillis() - timeAtLastReset > DRIFT_COOLDOWN_MS
+			) {
+				// Check and remove from lists to keep them under the reset
+				// limit
+				if (driftQuats.size() == driftQuats.capacity()) {
+					driftQuats.removeLast();
+					driftTimes.removeLast();
+				}
+
+				// Add new drift quaternion
+				driftQuats
+					.add(
+						new Quaternion()
+							.fromAngles(
+								0f,
+								rotQuat.mult(beforeQuat.inverse()).getYaw(),
+								0f
+							)
+					);
+
+				// Add drift time to total
+				driftTimes.add(System.currentTimeMillis() - driftSince);
+				totalDriftTime = 0;
+				for (Long time : driftTimes) {
+					totalDriftTime += time;
+				}
+
+				// Calculate drift Quaternions' weights
+				driftWeights.clear();
+				for (Long time : driftTimes) {
+					driftWeights.add(((float) time) / ((float) totalDriftTime));
+				}
+				// Make it so recent Quaternions weight more
+				for (int i = driftWeights.size() - 1; i > 0; i--) {
+					driftWeights
+						.set(
+							i,
+							driftWeights.get(i) + (driftWeights.get(i - 1) / driftWeights.size())
+						);
+					driftWeights
+						.set(
+							i - 1,
+							driftWeights.get(i - 1)
+								- (driftWeights.get(i - 1) / driftWeights.size())
+						);
```

Comment:
Can you explain what's going on here? This is so difficult to understand

Reply:
it's making it so the most recent data in the list is worth more.
To be precise, if you have 10, 20 and 5, it'll do:
```
10 = 10 + (20 / 3)
20 = 20 - (20 / 3)
```
and repeat for each one up to the last.
```
13.33 = 13.33 + (5 / 3)
5 = 5 - (5 / 3)
```
`13.33` being the second element in the list (used to be 20)
`3` being the amount of numbers of the list.

Reply:
ah hm I see, is there any way you could write a comment explaining this a little better in the code?

Reply:
Sure